File: src/bin/0150-evaluate-reverse-polish-notation.py
# @leetup=info id=150 lang=python3 slug=evaluate-reverse-polish-notation



from typing import Optional

# @leetup=code
class Solution:
    def evalRPN(self, tokens: list[str]) -> int:
        s = []
        for a in tokens:
            if a not in "+-*/":
                s.append(int(a))
                continue

            a2 = s.pop()
            a1 = s.pop()
            if a == "+":
                s.append(a1 + a2)
            elif a == "-":
                s.append(a1 - a2)
            elif a == "*":
                s.append(a1 * a2)
            elif a == "/":
                # a3 = abs(a1) // abs(a2)
                # if a1 < 0 < a2 or a2 < 0 < a1:
                #     a3 = -a3
                # s.append(a3)
                s.append(int(float(a1) / a2))
        assert len(s) == 1
        return s[0]


# @leetup=code

File: src/bin/0150-evaluate-reverse-polish-notation.cpp
// @leetup=info id=150 lang=cpp slug=evaluate-reverse-polish-notation

#include <algorithm>
#include <cassert>
#include <functional>
#include <iostream>
#include <map>
#include <set>
#include <stack>
#include <tuple>
#include <vector>

// @leetup=code

class Solution {
 public:
  int evalRPN(std::vector<std::string>& tokens) {
    std::stack<int> stack;
    for (std::string& s : tokens) {
      std::function<int(int, int)> op;
      if (s == "+") {
        op = std::plus<int>();
      } else if (s == "-") {
        op = std::minus<int>();
      } else if (s == "*") {
        op = std::multiplies<int>();
      } else if (s == "/") {
        op = std::divides<int>();
      }
      if (op) {
        int a1 = stack.top();
        stack.pop();
        int a2 = stack.top();
        stack.pop();
        stack.push(op(a2, a1));
      } else {
        stack.push(std::stoi(s));
      }
    }
    assert(stack.size() == 1);
    return stack.top();
  }
};
// @leetup=code

int main() {
  std::cout << "Hello, world!" << std::endl;
  return 0;
}
